src/models/kinflated_gamma_poisson_model.py

```diff
@@ -0,0 +1,551 @@
+        u = self._gamma_poisson_pmf(
+            np.arange(n + 1, MAXIMUM_COMPUTATIONAL_FREQUENCY), alpha, beta
+        )
+        pmf_tail = ((1 - s) / (1 - t)) * u
+        # self._pmf[n] = kGP(n | alpha, beta, a_1, ..., a_k)
+        self._pmf = np.array([0.0] + list(a) + list(pmf_tail))
+
+    def _gamma_poisson_pmf(self, n, alpha, beta):
+        """PMF of the shifted Gamma-Poisson distribution.
+
+        This implementation makes use of the equivalence between the
+        Gamma-Poisson distribution with parameters (alpha, beta) and
+        the negative binomial distribution with parameters (p, r) =
+        (1 / (1 + beta), alpha).
+
+        Args:
+          n:  Value(s) at which the distribution is to be evaluated.
+            This can be a scalar or a numpy array.
+          alpha: Alpha parameter of the Gamma-Poisson distribution.
+          beta: Beta parameter of the Gamma-Poisson distribution.
+        Returns:
+          f(n | alpha, beta), where f is the Gamma-Poisson distribution.
+        """
+        # Note scipy.stats.nbinom uses a different parameterization than
+        # the Wikipedia parameterization.  Under this parameterization,
+        # p -> 1-p, so the parameter 1/(1 + beta) becomes beta/(1 + beta).
+        return scipy.stats.nbinom.pmf(n - 1, alpha, 1.0 / (1.0 + beta))
+
+    def pmf(self, n):
+        """PMF of the k-inflated Gamma-Poisson distribution.
+
+        Args:
+          n:  Value(s) at which the distribution is to be evaluated.
+            This can be a scalar or a numpy array.
+        Returns:
+          kGP(n | alpha, beta, a_1, ..., a_k), where f is the k-inflated
+          Gamma-Poisson distribution.
+        """
+        return self._pmf[n]
+
+    def knreach(self, k, n, p):
+        """Probability that a random user has n impressions of which k are shown.
+
+        Computes the function:
+
+            C(n, k) p^k (1 - p)^{n-k} kGP(n | alpha, beta, a).
+
+        where C(n, k) is the binomial coefficient n!/[k!(n-k)!].
+
+        Args:
+          k:  Number of impressions that were seen by the user.
+          n:  Available inventory for the user.
+            This can be either a scalar or a numpy array.
+          p:  Probability that a given impression will be chosen.
+        Returns:
+          Probability that a randomly chosen user will have an inventory
+          of n impressions, of which k are shown.
+        """
+        return scipy.stats.binom.pmf(k, n, p) * self.pmf(n)
+
+    def kreach(self, k, p):
+        """Probability that a random user receives k impressions.
+
+        Computes the function:
+            sum_{n=k}^{infinity} g(k, n | alpha, beta, I, I_max).
+        where g(k, n | alpha, beta, I, I_max) is the probability that
+        a randomly chosen user will have an available inventory of n
+        impressions, of which k are shown.
+
+        Args:
+          k:  np.array specifying number of impressions that were seen by the user.
```

Comment:
Smart to allow k to be array here and n to be array above!

Reply:
Thanks!